controllers/core/event_controller.go

```diff
@@ -157,44 +199,82 @@ func (r *EventReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl
 		if !event.GetCreationTimestamp().After(time.Now().Add(-time.Minute*EventCreatedPastMinutes)) || event.Regarding.Kind != EventRegardingKind {
 			continue
 		}
-		if r.isValidEventForSGP(event) || r.isValidEventForCustomNetworking(event) {
+		eventForSGP := r.isValidEventForSGP(event)
+		eventForCN := r.isValidEventForCustomNetworking(event)
+		if eventForSGP || eventForCN {
 			nodeName := event.Regarding.Name
+			// use instance ID as cache key since they are guaranteed regionally unique
+			hostID := string(event.Regarding.UID)
+
+			// index 0 is the flag for SecurityGroupForPod, index 1 is the flag for Custom Networking
+			cacheValue := []byte{0, 0}
+			// changing to 1 indicates the event for the regarding feature
+			if eventForSGP {
+				cacheValue[EnableSGP] = 1
+			} else {
+				cacheValue[EnableCN] = 1
+			}
+
+			// use cache to avoid unnecessary calls to API server/ClientCache and node controller cache
+			// due to using List, the redundant call number can be large
+			if hitValue, cacheErr := r.cache.Get(hostID); cacheErr == nil {
+				if (eventForSGP && hitValue[EnableSGP] == 1) || (eventForCN && hitValue[EnableCN] == 1) {
+					eventControllerEventNodeCacheOpsCount.WithLabelValues([]string{CacheHitMetricsValue, "", ""}...).Inc()
+					logger.V(1).Info("Node has been processed and found in cache, will skip this event", "NodeName", nodeName, "InstanceID", hostID)
+					continue
+				}
+			}
+			// we use the GetNode() as a safeguard to avoid repeating reconciling on deleted nodes
 			if node, err := r.K8sAPI.GetNode(nodeName); err != nil {
 				// if not found the node, we don't requeue and just wait VPC CNI to send another request
 				r.Log.V(1).Info("Event reconciler didn't find the node and wait for next request for the node", "Node", node.Name)
 				return ctrl.Result{}, nil
 			} else {
-				if r.isEventToManageNode(event) && r.isValidEventForSGP(event) {
+				eventControllerEventNodeCacheOpsCount.WithLabelValues([]string{"", CacheMissMetricsValue, ""}...).Inc()
+				eventControllerEventNodeCacheSize.Set(float64(r.cache.Len()))
+				if r.isEventToManageNode(event) && eventForSGP {
 					// make the label value to false that indicates the node is ok to be proceeded by providers
 					// provider will decide if the node can be attached with trunk interface
 					labelled, err := r.K8sAPI.AddLabelToManageNode(node, config.HasTrunkAttachedLabel, config.BooleanFalse)
```

Comment:
Nit: Maybe later we can move this logic into a small utility function to label the node and increment the counter.

Reply:
Makes sense. I can follow up with this in another PR. Thanks.